### PR TITLE
Fluid Garbage Collection - Draft

### PR DESCRIPTION
**Pull request by agarwal-navin**
This is a prototype for garbage collection in fluid. It's in a very early stage and lot of things are very draft-y.
An overview of what this change does:
1. Container runtime runs garbage collection in summaryInternal after it has summarized each data store. It then adds the data stores that are to be deleted in a separate blob in the main summary.
**Note:** This will change and mostly be part of each individual data store's tree.
2. Container runtime requests the referenced data stores for a list of routes to referenced fluid objects in it.
3. Data store in turn requests each DDS for a list of routes to referenced fluid objects in it.
4. Each DDS keeps track of the route to any fluid object in it. It does so during snapshot by capturing the absolutePath to each IFluidHandle it serializes. I made a couple of changes for this:
  4.1. Changed snapshot method in SharedObject to not be abstract. It now calls into the FluidSerializer and asks it to record each serialized handle's absolutePath during the snapshot process.
  4.2. SharedObject gets this list from the FluidSerializer and caches it.
  4.3. Each DDS now implements snapshotCore which does exactly the same thing as snapshot does today. It is called by SharedObject's snapshot method.
  4.4. When asked for the list of routes to referenced fluid objects, the SharedObject returns the list cached above.
  4.5. At any given point in time, this list corresponds to the data in the last snapshot.
**Note:** I am currently using the FluidSerializer of the DataStoreRuntime to record the list of serialized handles and store the information in the SharedObject. Another implementation I was thinking of was to have each SharedObject its own FluidSerializer and store the information there.

### CODE REVIEW

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 1]
```diff
@@ -186,7 +192,17 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
     /**
      * {@inheritDoc (ISharedObject:interface).snapshot}
      */
-    public abstract snapshot(): ITree;
+    public snapshot(): ITree {
+        this.runtime.IFluidSerializer.startRecordingRoutes();
```
Comment (agarwal-navin) [inline packages/dds/shared-object-base/src/sharedObject.ts:196]: It might be worth for each SharedObject to have its own FluidSerializer. The serializer can then store the referenced routes per DDS.

Comment (agarwal-navin) [reply packages/dds/shared-object-base/src/sharedObject.ts:196]: Another option can be is to have a 'serialize' function in the SharedObject that DDSs call to serialize handles. This 'serialize` method stores the 'path' of the serialized handle. This way we don't have to modify the serializer.

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 2]
```diff
@@ -61,6 +61,12 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
      */
     private _isBoundToContext: boolean = false;
 
+    private _referencedRoutes: string[] = [];
```
Comment (vladsud) [inline packages/dds/shared-object-base/src/sharedObject.ts:64]: It feels like it should be a set, not an array

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 1 above)
Comment (vladsud) [reply packages/dds/shared-object-base/src/sharedObject.ts:196]: Yes for new serialize. It's trivial to make it. I just moved it from container to data store layer, but we can just make them at DDS level. I'd recreated them on demand (for summary) instead of adding some APIs to capture - that would not work before my recent change because I believe we summarize data stores in parallel (so you are getting lucky here with timing :))

Comment (vladsud) [reply packages/dds/shared-object-base/src/sharedObject.ts:196]: I'd go even further and explore passing serializer into snapshot() API and down, to be more explicit.
That has a risk though, as code may accidently use instance on data store. So not sure

File: packages/loader/container-loader/src/deltaManager.ts [hunk 3]
```diff
@@ -1213,7 +1213,7 @@ export class DeltaManager
                         const error = {
                             errorType: ContainerErrorType.dataCorruption,
                             message: "Two messages with same seq# and different payload!",
-                            clientId: this.connection?.details.clientId,
```
Comment (vladsud) [inline packages/loader/container-loader/src/deltaManager.ts:1216]: I fixed it :) Pull from main

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 4]
```diff
@@ -784,7 +789,7 @@ export class ContainerRuntime extends EventEmitter
                 }
                 const snapshotTree = value as ISnapshotTree;
                 // Need to rip through snapshot.
-                const { pkg, snapshotFormatVersion } = readAndParseFromBlobs<IFluidDataStoreAttributes>(
+                const { pkg, snapshotFormatVersion, isRootStore } = readAndParseFromBlobs<IFluidDataStoreAttributes>(
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:792]: Consider this PR as a prototype, but once you are ready to commit things, we may want to go with multiple smaller PRs.
One addressing differentiation of root vs. non-root data stores would be nice to have in separate PR.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 5]
```diff
@@ -1466,11 +1478,126 @@ export class ContainerRuntime extends EventEmitter
                 builder.addWithStats(key, contextSummary);
             }));
 
+        const deletedDataStores = this.runGarbageCollection();
+        builder.addBlob(".deletedDataStores", JSON.stringify(deletedDataStores));
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1482]: We should figure out proper mark up here, not put it into the blob. Storage is not going to read blobs.
Also any solution here should assume deletes on any level in the tree. Consider future of data stores being on channels, i.e. data store being a child of data store - the design that we put in place here should work with minimal differences for such case (DDSs is a another case that we need to make sure can be supported in future)

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 6]
```diff
@@ -1466,11 +1478,126 @@ export class ContainerRuntime extends EventEmitter
...
+
         this.serializeContainerBlobs(builder);
         const summary = builder.getSummaryTree();
         return { ...summary, id: "" };
     }
 
+    public runGarbageCollection(): string[] {
+        // Maintains a map of dataStoreId to an IReferences object. The IReferences object keeps track of whether
+        // the data store has been visited (to detect loops) and the fluid objects in it that are referenced.
+        const dataStoreReferencesMap: Map<string, IReferences> = new Map();
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1492]: FYI: This should be a path in the future, i.e. string[]. to account for nested data stores. string might be fine for now, but if we can keep things generic with clear comments on why, then it would be better.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 7]
```diff
@@ -1466,11 +1478,126 @@ export class ContainerRuntime extends EventEmitter
...
+
+        // A stack of routes that are referenced through handles in DDSes.
+        const referencedRoutesStack: string[] = [];
+
+        // Start with the root data stores. Add all the fluid objects referenced by the root data stores in to the
+        // referencedObjectsStack.
+        this.contexts.forEach((value: FluidDataStoreContext) => {
+            if (value.isRootStore) {
+                // Ignore the scheduler.
+                if (value.id === taskSchedulerId) {
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1502]: why task scheduler is special?

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 8]
```diff
@@ -1466,11 +1478,126 @@ export class ContainerRuntime extends EventEmitter
...
+                // Get the fluid objects referenced by this data store and add them to the referencedObjectsStack.
+                const referencedRoutes = value.getReferencedRoutes();
+                referencedRoutesStack.push(...referencedRoutes);
+
+                // Set the IReferences object in the dataStoreReferencesMap for this data store. Mark it as visited.
+                dataStoreReferencesMap.set(value.id, { visited: true, referencedRoutes: new Set<string>() });
+            }
+        });
+
+        // Do a Bread First Search on the referencedRoutesStack.
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1515]: bread -> breadth

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 9]
```diff
@@ -1466,11 +1478,126 @@ export class ContainerRuntime extends EventEmitter
...
+        while (referencedRoutesStack.length > 0) {
+            // Get the next route.
+            const nextRoute = referencedRoutesStack.shift();
+            assert(nextRoute);
+
+            // Get the dataStoreId from the referenced object.
+            const dataStoreId = nextRoute.split("/")[1];
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1522]: Is there an enforcement that it always starts with /?

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 10]
```diff
@@ -1466,11 +1478,126 @@ export class ContainerRuntime extends EventEmitter
...
+            }
+
+            const visited: boolean = references.visited;
+            // Keeps track of whether the entry for this data store in the dataStoreReferencesMap changes.
+            let dataStoreEntryChanged: boolean = false;
+
+            // Get the path of the referenced object relative to the dataStoreId and add it to the
+            // referencedFluidObjects in its references.
+            const dataStoreIdRelativePath = nextRoute.slice(dataStoreId.length + 1);
+            if (dataStoreIdRelativePath !== "") {
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1540]: I think (but not sure) that this results in never adding "/" to the list, i.e. path that does not go deeper (default path)

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 11]
```diff
@@ -1466,11 +1478,126 @@ export class ContainerRuntime extends EventEmitter
...
+                assert(references.referencedRoutes);
+                references.referencedRoutes.add(dataStoreIdRelativePath);
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1542]: Something to keep in mind: With current approach, we will never detect cycles involving DDSs. I.e. if data stores are references, but their DDSs create unreferenced (from outside) cycle, then they will be considered referenced and never collected.
So any future potential to collect DDSs would require substantially different algorithm. Maybe we from the start say that DDS collection is out of the window.
That all said, giving info to data store on unused (or used) paths is still valuable, as (IMHO) we have to figure out a way for data stored to collect sub-DDS objects, like at-mention where all at-mentions are part of same DDS. This requires file change (i.e. ops on the wire), but it's something that we will need to figure out in the future.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 12]
```diff
@@ -1466,11 +1478,126 @@ export class ContainerRuntime extends EventEmitter
...
+            }
+
+            // If this data store has not been visited yet, mark it as visited.
+            if (!references.visited) {
+                references.visited = true;
+                dataStoreEntryChanged = true;
+            }
+
+            // If the entry has changed, update it in the dataStoreReferencesMap.
+            if (dataStoreEntryChanged) {
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1553]: Can this update be unconditional (and drop dataStoreEntryChanged variable)?

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 13]
```diff
@@ -1466,11 +1478,126 @@ export class ContainerRuntime extends EventEmitter
...
+
+        const deletedDataStores: string[] = [];
+        // Now we have a map of all the data stores that are referenced and the routes that are referenced in it.
+        // Iterate through all the data stores in this runtime and:
+        // 1. If they are not referneced, mark them for collection.
+        // 2. If they are referenced, pass the list of routes in them that are referenced so it can collect
+        //    unreferenced objects.
+        this.contexts.forEach((value: FluidDataStoreContext) => {
+            // Ignore the scheduler.
+            if (value.id === taskSchedulerId) {
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1575]: same question here

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 14]
```diff
@@ -1466,11 +1478,126 @@ export class ContainerRuntime extends EventEmitter
...
+            // If the data store does exists in the dataStoreReferencesMap, it is no referenced. Mark it for collection.
+            // Else, get the list of the referenced routes in it and ask it to collect the unreferenced objects.
+            if (!dataStoreReferencesMap.has(value.id)) {
+                deletedDataStores.push(value.id);
+                value.setReferenced(false);
+            } else {
+                const references = dataStoreReferencesMap.get(value.id);
+                assert(references);
+                value.setReferenced(true);
+                value.collectUnreferencedFluidObjects([...references.referencedRoutes]);
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1588]: I'd change name to notifyUsedRoutes() or something like that, otherwise name and argument are at odds with each other

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 15]
```diff
@@ -1466,11 +1478,126 @@ export class ContainerRuntime extends EventEmitter
...
+            }
+        });
+        return deletedDataStores;
+    }
+
+    private getDataStoreReferencedRoutes(dataStoreId: string): string[] {
+        const dataStore = this.contexts.get(dataStoreId);
+        assert(dataStore);
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1596]: This assumes that ALL data stores are loaded, right?
I think that can't be the case, we need pay-per-play solution where data store is loaded only if it was changed, or maybe for legacy files where we do not have info from last summary stored in snapshot about routes used in it.

File: packages/runtime/container-runtime/src/dataStoreContext.ts [hunk 16]
```diff
@@ -375,12 +390,26 @@ export abstract class FluidDataStoreContext extends EventEmitter implements
...
     }
 
+    public getReferencedRoutes(): string[] {
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        return this.channel!.getReferencedRoutes();
+    }
+
+    public collectUnreferencedFluidObjects(referencedObjectRoutes: string[]) {
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        this.channel!.collectUnreferencedFluidObjects(referencedObjectRoutes);
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/dataStoreContext.ts:406]: If data store was not loaded (i.e. prior results were used), I assume we will not try to load it to clean it, right?

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 17]
```diff
@@ -567,10 +567,21 @@ export class FluidDataStoreRuntime extends EventEmitter implements IFluidDataSto
...
                 return new TreeTreeEntry(key, snapshot);
             }));
-
         return entries;
     }
 
+    public getReferencedRoutes(): string[] {
+        const referencedRoutes: string[] = [];
+        this.contexts.forEach((value: IChannelContext) => {
+            const routes = value.getReferencedRoutes();
```
Comment (vladsud) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:576]: same feedback as in container - we need to use prior results (from prior snapshot) if DDS does not need to be loaded because it did not change.
We have sort-of chicken-and-egg problem. On one hand, we need to run summarizer logic as it knows if certain things actually changed. But on order to summarize, we need "is it referenced" bit (to be written into summary), which requires summarizing logic to run.

**Review by vladsud — COMMENTED** [review]
FYI - I did not examine all the logic carefully, would need another pass. Please think about how to stage it properly, as adding all the missing pieces (especially pay-per-play) would balloon this change into really large, and really hard to review

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 2 above)
Comment (agarwal-navin) [reply packages/dds/shared-object-base/src/sharedObject.ts:64]: Yeah, set makes more sense. Eventually GC uses set so it will filter duplicates but its a good idea to do it at this level too.

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 1 above)
Comment (agarwal-navin) [reply packages/dds/shared-object-base/src/sharedObject.ts:196]: I had moved FluidSerilizer from Container to data store in this change too but your change beat me to it :-).
I like the idea of it being per-DDS level too because its not expensive at all and makes the code simpler.

I am not very comfortable with passing it to snapshot() because of the reasons you mentioned. It increases the surface area of things going wrong.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 4 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:792]: Yep, agreed. I will need to break it down. I wanted to get an idea of how people feel about this prototype and if things look good, I will break it down into smaller pieces and check them in separately.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 5 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1482]: Yeah, I added it here for now to make my prototpye testable :-).
I spoke to Arin at length about where / how we should put this information. We came up with a couple of ideas and I am going to play around with them in my next prototype.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 6 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1492]: Okay, I would love to understand about nested data stores and what the timeline for that is. I think it make it easier for me to implement keeping that in mind.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 7 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1502]: I guess this is because of my lack of understanding of the scheduler. I thought we cannot put references to other fluid objects in the scheduler. Is that understanding not correct?

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 8 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1515]: lol :D

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 9 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1522]: Yes, I am using handle's absolutePath which always starts with a "/".

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 10 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1540]: We don't need to add it to the list right. The list keeps track of all child objects withing a data store that are referenced. So, we don't need to add "/" to the list.
It is already part of the "referencedRoutesStack" which keeps track of all referenced data stores.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 12 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1553]: I believe so. I added this flag as I didn't want to unnecessary update the map. I can remove this.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 15 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1596]: This is something that needs some thought and work. Currently, when you call a data store to get snapshot, its all or nothing - it will either take entire snapshot or not. We need a mechanism that says something like - the only thing that changed is whether its referenced or not, so only update that part of the summary tree.

File: packages/runtime/container-runtime/src/dataStoreContext.ts (hunk 16 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/dataStoreContext.ts:406]: Yes, we should never load anything unnecessarily.

File: packages/runtime/datastore/src/dataStoreRuntime.ts (hunk 17 above)
Comment (agarwal-navin) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:576]: Exactly! Like I mentioned in some of previous comments, I am trying to work out a solution here that requires minimum work and is simple from a design point of view.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 7 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1502]: Even if it's the case, it does not matter - it will return empty set. We should not make any exclusions like that, this allows system to be flexible, i.e. there is no need to put restrictions around something that may want not to have such restrictions in the future

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 10 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1540]: / is also a child object (in current system). I.e. I can have reference of type /ID/ and that refers to default object of data store - data store needs to know about it.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 12 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1553]: If update is a single property, then it's cheaper to do it then to do many writes to this variable and have IF statement. branching is the most expensive stuff for processor :)

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 15 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1596]: No, it's not all or nothing. Data store returns snapshot, but it can reuse bits and pieces (sub-trees) from previous snapshot. So it's pay-per-play on every level. Yes, it eventually looks like full snapshot, but storage make it a reality by flattening it (i.e. resolving old references).
That said, we should not load data store at all if it did not change - that's the issue I'm raising here.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 10 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1540]: Ah okay. I will add code to handle this.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 12 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1553]: True, branching is expensive :-).
